src/uu/date/src/date.rs

```diff
@@ -219,19 +219,27 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
                 let iter = std::iter::once(date);
                 Box::new(iter)
             }
-            DateSource::File(ref path) => match File::open(path) {
-                Ok(file) => {
-                    let lines = BufReader::new(file).lines();
-                    let iter = lines.filter_map(Result::ok).map(parse_date);
-                    Box::new(iter)
-                }
-                Err(_err) => {
+            DateSource::File(ref path) => {
+                if path.is_dir() {
                     return Err(USimpleError::new(
                         2,
-                        format!("{}: No such file or directory", path.display()),
+                        format!("expected file, got directory"),
                     ));
                 }
-            },
+                match File::open(path) {
+                    Ok(file) => {
+                        let lines = BufReader::new(file).lines();
+                        let iter = lines.filter_map(Result::ok).map(parse_date);
+                        Box::new(iter)
+                    }
+                    Err(_err) => {
+                        return Err(USimpleError::new(
+                            2,
+                            format!("{}: No such file or directory", path.display()),
```

Comment:
It seems like this is not always the no such file or directory error, right? Can we use some of the [`UIoError`](https://docs.rs/uucore/latest/uucore/error/struct.UIoError.html) machinery to make this support all io errors?

Reply:
This [commit](https://github.com/uutils/coreutils/pull/4572/commits/20c24433ac1eb10a69fce2e6706056abd7b70ba3) should fix it.

Reply:
Sorry I see now that it was unrelated to your change. And this actually is a bit complicated because of the exit code of 2. Might be best to keep it like it was after all.

Reply:
I don't think you were wrong. The error code is being mapped from `std::io::Error` to `UIoError` which is producing correct error messages for different cases.

In my testing, I ran `date` on a file with no read permissions and it returned "permission denied" error. Next, I tried giving a non-existent file path and it returned "no such file or directory" error.

It seems to be working as intended.

Reply:
Yes, but the exit code is one now right? A test for it is also failing because you've changed the format of the error slightly:
```
 run: /home/runner/work/coreutils/coreutils/target/debug/coreutils date --file invalid_file
thread 'test_date::test_date_for_invalid_file' panicked at 'assertion failed: `(left == right)`
  left: `"date: 'invalid_file': No such file or directory"`,
 right: `"date: invalid_file: No such file or directory"`', tests/by-util/test_date.rs:280:5
 ```

Reply:
It's on separate commits right? You can do an interactive rebase without those commits and then force push

Reply:
@tertsdiepraam Yes and I plan on doing the very same thing. I was wondering if I update the error message to "Unknown error ocurred".

```sh
$ target/debug/coreutils date -f /tmp/foo
date: /tmp/foo: Unknown error ocurred  # The file exists but user doesn't have read access.
```

A generic error is warranted here, isn't it?

Reply:
I'm not sure. The general compatibility guarantee we try to follow is roughly "if GNU succeeds then uutils should too", so I think we actually want to be quiet? It's up for debate though. A specific error is always then a generic one. We could of course print the first io error that occured, instead of failing silently.

At least for this PR, let's follow GNU and add a test and we can discuss other changes separately.

Reply:
@tertsdiepraam is this okay?